---
title: "Copy number and chromosome 1q abnormality in multiple myeloma in single-cell whole genome sequencing "
output: html_document
date: "2024-11-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# How to use HMMCopyClusterPlody R modules to build a hierarchycal clustering models to estimate tumor clones and examine copy number abnormality (CNA).

## HMMCopyClusterPlody: HMMCopy hierarchical clustering and ploidy explorer in multiple myeloma (MM) single cell whole genome sequencing (scWGS)

• This is R package, HMMCopyClusterPloidy v1.2.

• The current version of HMMCopyClusterPlody is beta 0.0.1 (November 27th, 2024).

## A. Introduction

• Here we developed a module to analyze copy number abnormality (CNA) data generated by HMMCopy in scWGS data of multiple myleoma patients. HMMCopyClusterPlody module will generate hierchical clustering heatmap to estimate tumor clones and ploidy plot to represent hyperploidy in MM patients.

## B. Install R package

```{Installation}
install.packages("devtools")
library(devtools)

devtools::install_github("sanghoonleepitt/iGenSigRx")
library(HMMCopyClusterPloidyv1.2)
```

## B. Preprocess HMMCopy CNA state data 

### Read the HMMcopyRead.csv.gz file and process it.

  - Remove outlier bins that have NA values > 20 % of cells 
  - Remove artifact cells with Gini coefficient > median(GiniCoefficient)+ mad(GiniCoefficient)


```{Preprecessing}
####### ++++++++++++  Patient 3
HMMcopyReadFile <- "/Users/lees130/Library/CloudStorage/OneDrive-NYULangoneHealth/N07_DLPplus_scWGS/04_DLPplus_1qGain_8ptData/01_206_143839A/hmmcopy_reads.csv.gz"  
MySampleID="01_206_143839A"

HMMcopy_AfterRmvCell_WithoutNABinRegion <- CNVStateProcess(HMMcopyReadFile,RemoveArtifactCell=TRUE, BinRegionDelete=TRUE, MySampleID)
dim(HMMcopy_AfterRmvCell_WithoutNABinRegion) #  656 5228

```

## Step2. Make a hierarchical clustering heatmap for all chromosomes.

```{Hiearchical cluatering}
# HierarchialClusteringHeatmap(Processed_HMMcopyStateData=HMMcopy_AfterRmvCell_WithoutNABinRegion, MySampleID="01_206_143839A", AllCell="WithoutArtifact", ChromosomeToSubset="All",ClusterSize=9)

```


## Step3. Make a ploidy plot for all chromosomes.

```{Hiearchical cluatering}
### Patient #3 - 01-206_143839A
CNVSegmentFile=paste0(DLPpHMMCopyOutDir,"/01_206_143839A/hmmcopy_segments.csv.gz");
HierarchialClusterInfoFile=paste0(VignetteDirectory, "/RowSplitHeatmapDraw_WithoutArtifact_AllChr_01_206_143839A.rds");
HierarchialClusterInfoAfRmvClusterFile=paste0(VignetteDirectory, "/RowSplitHeatmapDraw_WithoutArtifactByClusterLowHighState2_AllChr_01_206_143839A.rds");
# CNVStateData_AfterRmvHighGiniMeanFile=paste0(HMMCopy_AfterRmvCellByGiniMean_Dir, "/HMMcopy_StateIdeal_RmvCellByGiniMean_WithoutBinRegion_01_206_143839A.rds");
CNVStateData_AfterRmvHighGiniMeanFile=paste0(VignetteDirectory, "/HMMcopy_StateIdeal_RmvCellByGiniMeanByCluster_WithoutBinRegion_01_206_143839A_LowState0.2.rds");
ClusterNumbToRmv=NULL;  ClusterNumbToPick=c(1:2); ClusterNumbToPick_InChr1 <- c(1:2); MySampleID="01_206_143839A"; 
HierarchialClusterInfoAfRmvClusterChr1File <- NULL  

Karyoplote_BySegment(CNVSegmentFile=CNVSegmentFile,HierarchialClusterInfoFile=HierarchialClusterInfoFile,
                     HierarchialClusterInfoAfRmvClusterFile=HierarchialClusterInfoAfRmvClusterFile,HierarchialClusterInfoAfRmvClusterChr1File=HierarchialClusterInfoAfRmvClusterChr1File,
                     CNVStateData_AfterRmvHighGiniMeanFile=CNVStateData_AfterRmvHighGiniMeanFile,
                     CNVStateData_AfterRmvHighGiniMeanClusterFile=CNVStateData_AfterRmvHighGiniMeanClusterFile,
                     ClusterNumbToRmv=ClusterNumbToRmv, ClusterNumbToPick=ClusterNumbToPick, ClusterNumbToPick_InChr1=ClusterNumbToPick_InChr1, MySampleID=MySampleID )

```
